### PR TITLE
fixes #3603: replaced a relation with a collection in 'where' conditions

### PR DESCRIPTION
**Pull request by dmitri-d**

### CODE REVIEW

**Comment by domcleal** [comment]
Could you add some tests please?

File: app/controllers/puppetclasses_controller.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def index
       values = Puppetclass.search_for ""
     end
     @puppetclasses = values.paginate(:page => params[:page])
-    @host_counter = Host.group(:puppetclass_id).joins(:puppetclasses).where(:puppetclasses => {:id => @puppetclasses.pluck(:id)}).count
+    @host_counter = Host.group(:puppetclass_id).joins(:puppetclasses).where(:puppetclasses => {:id => @puppetclasses.collect(&:id)}).count
```
Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb:17]: IMHO, if we are changing it already, we might as well change it to an INNER SELECT style which rails 3.2+ supports.

this was done prior to that.

Comment (dmitri-d) [inline app/controllers/puppetclasses_controller.rb:17]: no point, as @puppetclasses is used in the view.

Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb:17]: not sure i follow, afair, you can just change

``` ruby
where(:puppetclasses => @puppetclasses)
```

Comment (dmitri-d) [inline app/controllers/puppetclasses_controller.rb:17]: Nods, I suppose that is more concise. I'll see if sub select works with this syntax (the original issue was AR stripping all joins from the relation).

**Comment by dmitri-d** [comment]
@domcleal: now with tests.
@ohadlevy: same issue as before -- it looks like subqueries get stripped of joins (probably all parts but "where" predicates get stripped)

**Comment by domcleal** [comment]
Merged as 9e08f306 for Foreman 1.3.2, thanks @witlessbird.

**Pull request closed without merging**